File: Numeros Pseudo-Aleatorios/Middle Square/MiddleSquare.scala
/*
 PRNG .......: Middle Square
 Ano ........: 1946
 Autor ......: John von Neumann
*/

object Main {
  def main(args: Array[String]): Unit = {

    import scala.util.Random

    def meanSquare(seed: Long): Long = {
      val updatedSeed = (seed * seed / 100) % 10000
      if (updatedSeed < 0) updatedSeed + 10000 else updatedSeed
    }

    var Xn1: Long = 1234
    var Xn2: Long = System.currentTimeMillis() % 10000

    println("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)")
    println("PSEUDORANDOM NUMBER GENERATOR (PRNG)")
    println("PADRAO: Middle Square")
    println("==========================================\n")
    println("    Semente: 1234        Semente: TIME")
    println("    -------------        -------------\n")

    for (_ <- 1 to 10) {
      Xn1 = meanSquare(Xn1)
      Xn2 = meanSquare(Xn2)
      println(f"${Xn1}%17d        ${Xn2}%13d")
    }
  }
}

File: Numeros Pseudo-Aleatorios/Middle Square/MiddleSquare.c
/*
 PRNG .......: Middle Square
 Ano ........: 1946
 Autor ......: John von Neumann
*/

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

long int MeanSquare(long int *seed) {
    *seed = (*seed * *seed / 100) % 10000;
    return *seed >= 0 ? *seed : -*seed;
}

int main(void) {

    long int Xn1, Xn2;
    int i;

    Xn1 = 1234;
    Xn2 = time(NULL);

    printf("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)\n");
    printf("PSEUDORANDOM NUMBER GENERATOR (PRNG)\n");
    printf("PADRAO: Middle Square\n");
    printf("==========================================\n\n");
    printf("    Semente: 1234        Semente: TIME\n");
    printf("    -------------        -------------\n");

    for (i = 1; i <= 10; i++) {
        printf("%17ld        %13ld\n", MeanSquare(&Xn1), MeanSquare(&Xn2));
    }

    return 0;
}
